features/323-login-id/design.md

```diff
@@ -0,0 +1,253 @@
+
+## GitHub
+1. Supports multiple emails per user, with email verification.
+2. User can add/delete any amount of emails.
+3. Only verified emails would be associated with the user in public.
+4. User can allow recovery email/security alert to be sent to all verified emails or a specific verified email.
+
+
+## Facebook
+5. User can login with email/phone number/username, without indicating what type of value it is.
+6. User can add/delete any amount of emails, which can be used to login after verification.
+7. Disposable email address is blocked.
+8. All emails/phones of user is verified.
+9. Email communications are personalized with user language & name.
+
+## Other systems
+10. User may choose to login as different roles (e.g. teacher/student, admin/member).
+
+
+# Design Objectives
+1. Allow auth related features to understand semantics of some login IDs
+2. Support multiple login ID with same login ID key
+3. Support compound login IDs
+4. Support adding/removing login ID
+
+
+# Proposed Design Overview
+
+## Use Implicit Login ID Key for Logging in
+In current design, user would login using tuple (login ID key, login ID). i.e:
+```json
+{
+    "login_id": {
+        "email": "a@example.com"
+    },
+    "password": "12345678"
+}
+```
+
+To support logging in without specifiying login ID key (implicit login ID key),
+API consumer would just need to pass the login ID to the API. i.e.
+```json
+{
+    "login_id": "a@example.com",
+    "password": "12345678"
+}
+```
+
+This addresses use case 5.
+
+
+## Login ID Realm
+If developer would like user to login with compound login ID, current suggested
+approach is to use serialized compound login ID as login ID value.
+i.e.:
+```json
+{ "login_id": "student:a@example.com" }
+{ "login_id": "teacher:a@example.com" }
+```
+
+This approach would not allow auth features to understand semantics of the 
+login ID component, since the format of serialization is left unspecified.
+
+To support compound login ID better, a Realm would be associated with each
+password authenticated principal.
+- Realm is a user specified value, auth gear does not concern its semantics.
+- Default realm is 'default' if unspecified, for application without need of compound login ID.
+- Possible realm values are whitelisted.
+- Realm would be considered when matching login ID with principal.
+  i.e. password authenticated principals has unique (realm, login ID).
+- API consumer could specify realm along with login ID when login ID is required.
+
+This solution should support majority of use case for compound login ID.
+
+For example, when logging in with realms:
+```json
+{ "realm": "teacher", "login_id": "a@example.com", "password": "12345678" }
+{ "realm": "student", "login_id": "a@example.com", "password": "12345678" }
+```
+If authentication succeed, both request would result in same user, with
+different principals.
+
+This addresses use case 10.
+
+
+## Use Principals in Auth API
```

Comment:
I would suggest we either DO NOT introduce a new terms "Principle", or give a better definition here.

(apparently, the definition of principle is some data describing the current logged in session?)

Reply:
btw If my understanding of "Principle" here is correct -- the current logged in session, maybe give it a name that's more self explantory?